test/extended/prometheus/prometheus.go

```diff
@@ -49,6 +51,93 @@ type TelemeterClientConfig struct {
+			"openshift-operator-lifecycle-manager", // https://issues.redhat.com/browse/OCPBUGS-59768
+			"openshift-cluster-samples-operator",   // https://issues.redhat.com/browse/OCPBUGS-59769
+			"openshift-cluster-version",            // https://issues.redhat.com/browse/OCPBUGS-57585
+			"openshift-cluster-csi-drivers",        // https://issues.redhat.com/browse/OCPBUGS-60159
+		)
+	)
+
+	g.BeforeEach(func(ctx g.SpecContext) {
+		var err error
+		prometheusURL, err = helper.PrometheusRouteURL(ctx, oc)
+		o.Expect(err).NotTo(o.HaveOccurred(), "Get public url of prometheus")
+		bearerToken, err = helper.RequestPrometheusServiceAccountAPIToken(ctx, oc)
+		o.Expect(err).NotTo(o.HaveOccurred(), "Request prometheus service account API token")
+
+		if namespacesToSkip.Has(monitoringAuthTestNamespace) {
+			e2e.Logf("The namespace %s is not skipped because $MONITORING_AUTH_TEST_NAMESPACE is set to it", monitoringAuthTestNamespace)
+			namespacesToSkip.Delete(monitoringAuthTestNamespace)
+		}
+	})
+
+	g.It("should not be accessible without auth [Serial]", func() {
+		var errs []error
+
+		g.By("checking that targets reject the requests with 401 or 403")
+		execPod := exutil.CreateExecPodOrFail(oc.AdminKubeClient(), oc.Namespace(), "execpod-targets-authorization")
+		defer func() {
+			err := oc.AdminKubeClient().CoreV1().Pods(execPod.Namespace).Delete(context.Background(), execPod.Name, *metav1.NewDeleteOptions(1))
+			o.Expect(err).NotTo(o.HaveOccurred(), fmt.Sprintf("Delete pod %s/%s", execPod.Namespace, execPod.Name))
+		}()
+
+		contents, err := helper.GetURLWithToken(helper.MustJoinUrlPath(prometheusURL, "api/v1/targets"), bearerToken)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		targets := &prometheusTargets{}
+		err = json.Unmarshal([]byte(contents), targets)
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(len(targets.Data.ActiveTargets)).Should(o.BeNumerically(">=", 5))
+
+		expected := sets.New[int](http.StatusUnauthorized, http.StatusForbidden)
+		for _, target := range targets.Data.ActiveTargets {
+			ns := target.Labels["namespace"]
+			if !strings.HasPrefix(ns, "openshift-") {
+				continue
+			}
+			if monitoringAuthTestNamespace != "" && ns != monitoringAuthTestNamespace {
+				continue
+			}
+			pod := target.Labels["pod"]
+			e2e.Logf("Checking via pod exec status code from the scaple url %s for pod %s/%s without authorization (skip=%t)", target.ScrapeUrl, ns, pod, namespacesToSkip.Has(ns))
+			err := wait.PollUntilContextTimeout(context.Background(), 10*time.Second, 5*time.Minute, true, func(context.Context) (bool, error) {
+				statusCode, execError := helper.URLStatusCodeExecViaPod(execPod.Namespace, execPod.Name, target.ScrapeUrl)
+				e2e.Logf("The scaple url %s for pod %s/%s without authorization returned %d, %v (skip=%t)", target.ScrapeUrl, ns, pod, statusCode, execError, namespacesToSkip.Has(ns))
+				// execError is eaten up.
+				// Prometheus may have troubles to scrape the target, but that is out of the scope of this test
```

Comment:
As mentioned in other threads, if we cannot reach the targets, the test should fail, if this happens and
- if the test is broken, we should fix it.
- if the targets are not reachable, and retries didn't help to make sure auth is set up, the test should fail.

The test should fail if it cannot guarantee auth is set up.
If a target without auth is down for some reason and no other test fails because of that, this test should still fail, even if another test fails.

For all targets we should make sure we get 401 (or 403), for any other reason, the test should fail.

Reply:
Retry on errors too.
I was not sure what the contract between Prometheus and targets is.
Now it is clear that it has to be 401|403 without auth.
Neither runtime error nor 500 is accepted from your point of view.

I will create bugs for those cases (and add the namespaces to ignore as well).

Reply:
Done.

https://issues.redhat.com/browse/OCPBUGS-60258
https://issues.redhat.com/browse/OCPBUGS-60263